### PR TITLE
Fix the ordering of id freeing and deleting

### LINKED ISSUE #5372: Crashes with `Cannot remove a vacant resource` when running bevy on wgpu trunk.
**Description**
After running for anywhere from ~5-30 seconds, bevy examples are crashing at https://github.com/gfx-rs/wgpu/blob/b731495e053fe3a15879ff2637c2fb8db74ace3e/wgpu-core/src/storage.rs#L203 with `assertion left == right failed left: 1 right: 2`.

**Repro steps**
Run bevy's `3d_scene` example from my `wgpu-0.20` branch https://github.com/Elabajaba/bevy/tree/wgpu-0.20

**Expected vs observed behavior**
Expected: No crashes. Observed: crashes.

**Extra materials**
Relevant parts of the stack trace:
```
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::panicking::assert_failed_inner
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:342
   3: core::panicking::assert_failed<u32,u32>
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\panicking.rs:297
   4: wgpu_core::storage::Storage<wgpu_core::resource::StagingBuffer<wgpu_hal::vulkan::Api> >::remove<wgpu_core::resource::StagingBuffer<wgpu_hal::vulkan::Api> >
             at C:\Users\me\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\b731495\wgpu-core\src\storage.rs:203
   5: wgpu_core::registry::Registry<wgpu_core::resource::StagingBuffer<wgpu_hal::vulkan::Api> >::unregister<wgpu_core::resource::StagingBuffer<wgpu_hal::vulkan::Api> >
             at C:\Users\me\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\b731495\wgpu-core\src\registry.rs:157
   6: wgpu_core::global::Global::queue_write_staging_buffer<wgpu_hal::vulkan::Api>
             at C:\Users\me\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\b731495\wgpu-core\src\device\queue.rs:516
   7: wgpu::backend::wgpu_core::impl$7::queue_write_staging_buffer
             at C:\Users\me\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\b731495\wgpu\src\backend\wgpu_core.rs:2139
   8: wgpu::context::impl$5::queue_write_staging_buffer<wgpu::backend::wgpu_core::ContextWgpuCore>
             at C:\Users\me\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\b731495\wgpu\src\context.rs:2962
   9: wgpu::impl$74::drop
             at C:\Users\me\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\b731495\wgpu\src\lib.rs:4542
  10: core::ptr::drop_in_place<wgpu::QueueWriteBufferView>
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\ptr\mod.rs:507
  11: core::ptr::drop_in_place<bevy_render::render_resource::uniform_buffer::QueueWriteBufferViewWrapper>
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\ptr\mod.rs:507
  12: core::ptr::drop_in_place<encase::core::buffers::DynamicStorageBuffer<bevy_render::render_resource::uniform_buffer::QueueWriteBufferViewWrapper> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\ptr\mod.rs:507
  13: core::ptr::drop_in_place<encase::core::buffers::DynamicUniformBuffer<bevy_render::render_resource::uniform_buffer::QueueWriteBufferViewWrapper> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\ptr\mod.rs:507
  14: core::ptr::drop_in_place<bevy_render::render_resource::uniform_buffer::DynamicUniformBufferWriter<bevy_render::view::ViewUniform> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\ptr\mod.rs:507
  15: bevy_render::view::prepare_view_uniforms
             at .\crates\bevy_render\src\view\mod.rs:448
```

**Platform**
`wgpu` trunk b731495e053fe3a15879ff2637c2fb8db74ace3e
Windows 11
AdapterInfo { name: "AMD Radeon RX 6800 XT", vendor: 4098, device: 29631, device_type: DiscreteGpu, driver: "AMD proprietary driver", driver_info: "24.2.1 (AMD proprietary shader compiler)", backend: Vulkan }

edit: I've also confirmed it happens on windows amd 24.3.1 driver, Linux Mesa 24.0.3 AMD radv, and someone else has reproduced it on Windows + Nvidia 4080.

### PR DESCRIPTION
**Pull request by DJMcNab**
**Connections**
Fixes https://github.com/gfx-rs/wgpu/issues/5372

**Description**
When creating a resource in parallel with freeing a different resource on another thread, it was possible for the id to be made available before it was removed from storage.

That is, `unregister` was racing with e.g. `prepare().assign_existing()`

**Testing**
I cloned the linked reproduction, and applied this same patch.
The new version doesn't have this crash

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown` N/A
  - [ ] `--target wasm32-unknown-emscripten` N/A
- [x] Run `cargo xtask test` to run tests - no *new* failures
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

I'm not sure about a changelog entry, as this change is fixing a bug which was introduced in the 0.20 cycle

### CODE REVIEW

File: wgpu-core/src/registry.rs [hunk 1]
```diff
@@ -165,8 +165,8 @@ impl<T: Resource> Registry<T> {
         storage.insert_error(id, label);
     }
     pub(crate) fn unregister(&self, id: Id<T::Marker>) -> Option<Arc<T>> {
-        self.identity.free(id);
         let value = self.storage.write().remove(id);
+        self.identity.free(id);
```
Comment (cwfitzgerald) [inline wgpu-core/src/registry.rs]: Could you add documentation as to why this is in this order and documentation on the usage of assign_existing for why that's correct.

These bugs are super important to document well, as the interactions may be totally non-obvious to the outside (evidenced by us having the bug)

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Looks good, have docs/test comments. Thanks for investigating this!

Is it possible to write a test for this?

**Review by Elabajaba — APPROVED** [review]
Fix works for me!

**Review by cwfitzgerald — APPROVED** [review]
Good stuff

**Pull request merged**